opal/include/opal/sys/atomic_impl.h

```diff
@@ -34,182 +34,123 @@
  *
  * Some architectures do not provide support for the 64 bits
  * atomic operations. Until we find a better solution let's just
- * undefine all those functions if there is no 64 bit cmpset
+ * undefine all those functions if there is no 64 bit compare-exchange
  *
  *********************************************************************/
-#if OPAL_HAVE_ATOMIC_CMPSET_32
+#if OPAL_HAVE_ATOMIC_COMPARE_EXCHANGE_32
+
+#define OPAL_ATOMIC_DEFINE_CMPXCG_OP(type, bits, operand, name)  \
+    static inline type opal_atomic_ ## name ## _ ## bits (volatile type *addr, type value) \
+    {                                                                   \
+        type oldval, newval;                                            \
+        do {                                                            \
+            oldval = *addr;                                             \
```

Comment:
shouldn't this be outside the do loop ? (the oldval will be updated by the compare_exchange anyway).

Reply:
Ah yes. I made that change in other places but missed this one. Will be irrelevant in the next PR though. I will be replacing the underlying op-and-fetch arithmetic atomics with fetch-and-op. This is already fixed in that branch. I will post that PR once this one is merged unless you would like me to include the fetch-and-op change in this PR.